Namespaced scope fails for stress-ng plugin

## Describe the bug

When trying to use a namespaced scope with the stress-ng plugin ([development branch](https://github.com/arcalot/arcaflow-plugin-stressng/tree/simplify-schema), currently), the workflow will fail. The panic seems to randomly reference one of the sub-classes of the `stressors` one-of union in the input schema.

```
panic: Referenced object 'CpuStressorParams' not found in scope with namespace ""; available:
StressNGParams


goroutine 1 [running]:
go.flow.arcalot.io/pluginsdk/schema.(*RefSchema).ApplyScope(0x40005660c0, {0x182b030, 0x40002f4090}, {0x0, 0x0})
	/home/runner/go/pkg/mod/go.flow.arcalot.io/pluginsdk@v0.11.1/schema/ref.go:121 +0x214
go.flow.arcalot.io/pluginsdk/schema.OneOfSchema[...].ApplyScope(0x18379c0, {0x182b030, 0x40002f4090}, {0x0, 0x0})
	/home/runner/go/pkg/mod/go.flow.arcalot.io/pluginsdk@v0.11.1/schema/oneof.go:48 +0x98
go.flow.arcalot.io/pluginsdk/schema.AbstractListSchema[...].ApplyScope(...)
	/home/runner/go/pkg/mod/go.flow.arcalot.io/pluginsdk@v0.11.1/schema/list.go:85
go.flow.arcalot.io/pluginsdk/schema.(*PropertySchema).ApplyScope(...)
	/home/runner/go/pkg/mod/go.flow.arcalot.io/pluginsdk@v0.11.1/schema/property.go:126
go.flow.arcalot.io/pluginsdk/schema.(*ObjectSchema).ApplyScope(...)
	/home/runner/go/pkg/mod/go.flow.arcalot.io/pluginsdk@v0.11.1/schema/object.go:62
go.flow.arcalot.io/pluginsdk/schema.NewScopeSchema(0x40007eeb40, {0x0, 0x0, 0x2a?})
	/home/runner/go/pkg/mod/go.flow.arcalot.io/pluginsdk@v0.11.1/schema/scope.go:36 +0x234
go.flow.arcalot.io/engine/workflow.addScopesWithReferences(0x4000655200?, {0x182b030?, 0x40007a5aa0}, {0x4000382bd0, 0x2a})
	/home/runner/go/pkg/mod/go.flow.arcalot.io/engine@v0.16.0-beta1/workflow/executor.go:394 +0x18c
go.flow.arcalot.io/engine/workflow.addInputNamespacedScopes(0x40003f42d0?, {{{0x1538390, 0x7}, {0x1548700, 0x15}, {0x153a103, 0x9}, {0x15390f7, 0x8}, 0x40003417d0, ...}, ...}, ...)
	/home/runner/go/pkg/mod/go.flow.arcalot.io/engine@v0.16.0-beta1/workflow/executor.go:376 +0x1a0
go.flow.arcalot.io/engine/workflow.applyLifecycleScopes(0x4000567b40?, {0x182b030, 0x400000c930})
	/home/runner/go/pkg/mod/go.flow.arcalot.io/engine@v0.16.0-beta1/workflow/executor.go:326 +0x1f4
go.flow.arcalot.io/engine/workflow.(*executor).Prepare(0x4000567b40, 0x40000ff8b0, 0x40005afe30)
	/home/runner/go/pkg/mod/go.flow.arcalot.io/engine@v0.16.0-beta1/workflow/executor.go:124 +0x1e4
go.flow.arcalot.io/engine.workflowEngine.Parse({{0x1821ea0, 0x40004d70b0}, {0x1820690, 0x400058ff10}, 0x4000110660}, {0x1823c50, 0x40007a4558}, {0x1538cc7?, 0x400054daf8?})
	/home/runner/go/pkg/mod/go.flow.arcalot.io/engine@v0.16.0-beta1/engine.go:121 +0x28c
main.runWorkflow({0x180c098, 0x40004d78c0}, {0x1823c50, 0x40007a4558}, {0x1538cc7, 0x8}, {0x1821ea0?, 0x40004d7020}, {0x4000048e00, 0x6f8, ...})
	/home/runner/go/pkg/mod/go.flow.arcalot.io/engine@v0.16.0-beta1/cmd/arcaflow/main.go:204 +0x18c
main.main()
	/home/runner/go/pkg/mod/go.flow.arcalot.io/engine@v0.16.0-beta1/cmd/arcaflow/main.go:190 +0xb48
```

## To reproduce

The workflow here can be used to reproduce the problem: https://gitlab.com/redhat/edge/tests/perfscale/arcaflow-workflow-sysbench/-/blob/nightly-tests-stressng/stressng-workflow.yaml?ref_type=heads

Test with just the sub-workflow, un-commenting the `namespace` parameter for `$.inputs.StressngWorkflowInput.stressng_params`:
```
arcaflow -config config.yaml -input sample-input-stressng-cpu.yaml -context . -workflow stressng-workflow.yaml
```

## Additional context

The stress-ng plugin includes a one-of union of other classes with discriminators.

https://github.com/arcalot/arcaflow-plugin-stressng/blob/ebe8b964b3c8e3d37228585b09b27af0584f9778/arcaflow_plugin_stressng/stressng_schema.py#L551-L603

I'm not positive this is the source of the problem, but it seems suspect.

--- a/go.mod
+++ b/go.mod
@@ -13,7 +13,7 @@ require (
 	go.flow.arcalot.io/dockerdeployer v0.7.0
 	go.flow.arcalot.io/expressions v0.4.1
 	go.flow.arcalot.io/kubernetesdeployer v0.9.1
-	go.flow.arcalot.io/pluginsdk v0.11.1
+	go.flow.arcalot.io/pluginsdk v0.12.0
 	go.flow.arcalot.io/podmandeployer v0.11.0
 	go.flow.arcalot.io/pythondeployer v0.6.0
 	go.flow.arcalot.io/testdeployer v0.6.0

--- a/internal/util/invalid_serialization_detector.go
+++ b/internal/util/invalid_serialization_detector.go
@@ -19,6 +19,7 @@ func NewInvalidSerializationDetectorSchema() *InvalidSerializationDetectorSchema
 // the data has been serialized or unserialized at least once (since, if it is
 // never operated on, then it's trivial to claim that it was never doubly done).
 type InvalidSerializationDetectorSchema struct {
+	schema.ScalarType
 	SerializeCnt   int
 	UnserializeCnt int
 }
@@ -87,15 +88,6 @@ func (d *InvalidSerializationDetectorSchema) Serialize(data any) (any, error) {
 // SerializeType is string-input-typed version of Serialize.
 func (d *InvalidSerializationDetectorSchema) SerializeType(data string) (any, error) {
 	return d.Serialize(data)
-}
-
-// ApplyScope is for applying a scope to the references. Does not apply to this object.
-func (d *InvalidSerializationDetectorSchema) ApplyScope(_ schema.Scope, _ string) {}
-
-// ValidateReferences validates that all necessary references from scopes have been applied.
-// Does not apply to this object.
-func (d *InvalidSerializationDetectorSchema) ValidateReferences() error {
-	return nil
 }
 
 // TypeID returns the category of type this type is. Returns string because

--- a/workflow/any.go
+++ b/workflow/any.go
@@ -17,11 +17,7 @@ func newAnySchemaWithExpressions() *anySchemaWithExpressions {
 // anySchemaWithExpressions is a wildcard allowing maps, lists, integers, strings, bools, and floats. It also allows
 // expression objects.
 type anySchemaWithExpressions struct {
-	anySchema schema.AnySchema
-}
-
-func (a *anySchemaWithExpressions) ReflectedType() reflect.Type {
-	return a.anySchema.ReflectedType()
+	schema.AnySchema
 }
 
 func (a *anySchemaWithExpressions) Unserialize(data any) (any, error) {
@@ -39,23 +35,11 @@ func (a *anySchemaWithExpressions) ValidateCompatibility(dataOrType any) error {
 		// Assume okay
 		return nil
 	}
-	return a.anySchema.ValidateCompatibility(dataOrType)
+	return a.AnySchema.ValidateCompatibility(dataOrType)
 }
 
 func (a *anySchemaWithExpressions) Serialize(data any) (any, error) {
 	return a.checkAndConvert(data)
-}
-
-func (a *anySchemaWithExpressions) ApplyNamespace(objects map[string]*schema.ObjectSchema, namespace string) {
-	a.anySchema.ApplyNamespace(objects, namespace)
-}
-
-func (a *anySchemaWithExpressions) ValidateReferences() error {
-	return a.anySchema.ValidateReferences()
-}
-
-func (a *anySchemaWithExpressions) TypeID() schema.TypeID {
-	return a.anySchema.TypeID()
 }
 
 func (a *anySchemaWithExpressions) checkAndConvert(data any) (any, error) {
@@ -91,7 +75,7 @@ func (a *anySchemaWithExpressions) checkAndConvert(data any) (any, error) {
 	case reflect.String:
 		fallthrough
 	case reflect.Bool:
-		return a.anySchema.Unserialize(data)
+		return a.AnySchema.Unserialize(data)
 	case reflect.Slice:
 		result := make([]any, t.Len())
 		for i := 0; i < t.Len(); i++ {

--- a/workflow/any.go
+++ b/workflow/any.go
@@ -46,8 +46,8 @@ func (a *anySchemaWithExpressions) Serialize(data any) (any, error) {
 	return a.checkAndConvert(data)
 }
 
-func (a *anySchemaWithExpressions) ApplyScope(scope schema.Scope, namespace string) {
-	a.anySchema.ApplyScope(scope, namespace)
+func (a *anySchemaWithExpressions) ApplyNamespace(objects map[string]*schema.ObjectSchema, namespace string) {
+	a.anySchema.ApplyNamespace(objects, namespace)
 }
 
 func (a *anySchemaWithExpressions) ValidateReferences() error {

--- a/workflow/any.go
+++ b/workflow/any.go
@@ -48,34 +48,6 @@ func (a *anySchemaWithExpressions) checkAndConvert(data any) (any, error) {
 	}
 	t := reflect.ValueOf(data)
 	switch t.Kind() {
-	case reflect.Int:
-		fallthrough
-	case reflect.Uint:
-		fallthrough
-	case reflect.Int8:
-		fallthrough
-	case reflect.Uint8:
-		fallthrough
-	case reflect.Int16:
-		fallthrough
-	case reflect.Uint16:
-		fallthrough
-	case reflect.Int32:
-		fallthrough
-	case reflect.Uint32:
-		fallthrough
-	case reflect.Uint64:
-		fallthrough
-	case reflect.Int64:
-		fallthrough
-	case reflect.Float32:
-		fallthrough
-	case reflect.Float64:
-		fallthrough
-	case reflect.String:
-		fallthrough
-	case reflect.Bool:
-		return a.AnySchema.Unserialize(data)
 	case reflect.Slice:
 		result := make([]any, t.Len())
 		for i := 0; i < t.Len(); i++ {
@@ -102,8 +74,6 @@ func (a *anySchemaWithExpressions) checkAndConvert(data any) (any, error) {
 		}
 		return result, nil
 	default:
-		return nil, &schema.ConstraintError{
-			Message: fmt.Sprintf("unsupported data type for 'any' type: %T", data),
-		}
+		return a.AnySchema.Unserialize(data)
 	}
 }

--- a/workflow/executor.go
+++ b/workflow/executor.go
@@ -227,7 +227,7 @@ func (e *executor) processInput(workflow *Workflow) (schema.Scope, error) {
 	if !ok {
 		return nil, fmt.Errorf("bug: unserialized input is not a scope")
 	}
-	typedInput.ApplyScope(typedInput, schema.DEFAULT_NAMESPACE)
+	typedInput.ApplySelf()
 	return typedInput, nil
 }
 
@@ -317,7 +317,7 @@ func applyLifecycleScopes(
 	stepLifecycles map[string]step.Lifecycle[step.LifecycleStageWithSchema],
 	typedInput schema.Scope,
 ) error {
-	allNamespaces := make(map[string]schema.Scope)
+	allNamespaces := make(map[string]map[string]*schema.ObjectSchema)
 	for workflowStepID, stepLifecycle := range stepLifecycles {
 		for _, stage := range stepLifecycle.Stages {
 			prefix := "$.steps." + workflowStepID + "." + stage.ID + "."
@@ -335,9 +335,9 @@ func applyLifecycleScopes(
 // applyAllNamespaces applies all namespaces to the given scope.
 // This function also validates references, and lists the namespaces
 // and their objects in the event of a reference failure.
-func applyAllNamespaces(allNamespaces map[string]schema.Scope, scopeToApplyTo schema.Scope) error {
-	for namespace, scope := range allNamespaces {
-		scopeToApplyTo.ApplyScope(scope, namespace)
+func applyAllNamespaces(allNamespaces map[string]map[string]*schema.ObjectSchema, scopeToApplyTo schema.Scope) error {
+	for namespace, objects := range allNamespaces {
+		scopeToApplyTo.ApplyNamespace(objects, namespace)
 	}
 	err := scopeToApplyTo.ValidateReferences()
 	if err == nil {
@@ -345,9 +345,9 @@ func applyAllNamespaces(allNamespaces map[string]schema.Scope, scopeToApplyTo sc
 	}
 	// Now on the error path. Provide useful debug info.
 	availableObjects := ""
-	for namespace, scope := range allNamespaces {
+	for namespace, objects := range allNamespaces {
 		availableObjects += "\n\t" + namespace + ":"
-		for objectID := range scope.Objects() {
+		for objectID := range objects {
 			availableObjects += " " + objectID
 		}
 	}
@@ -359,13 +359,13 @@ func applyAllNamespaces(allNamespaces map[string]schema.Scope, scopeToApplyTo sc
 	)
 }
 
-func addOutputNamespacedScopes(allNamespaces map[string]schema.Scope, stage step.LifecycleStageWithSchema, prefix string) {
+func addOutputNamespacedScopes(allNamespaces map[string]map[string]*schema.ObjectSchema, stage step.LifecycleStageWithSchema, prefix string) {
 	for outputID, outputSchema := range stage.Outputs {
 		addScopesWithReferences(allNamespaces, outputSchema.Schema(), prefix+outputID)
 	}
 }
 
-func addInputNamespacedScopes(allNamespaces map[string]schema.Scope, stage step.LifecycleStageWithSchema, prefix string) {
+func addInputNamespacedScopes(allNamespaces map[string]map[string]*schema.ObjectSchema, stage step.LifecycleStageWithSchema, prefix string) {
 	for inputID, inputSchemaProperty := range stage.InputSchema {
 		var inputSchemaType = inputSchemaProperty.Type()
 		// Extract item values from lists (like for ForEach)
@@ -379,19 +379,19 @@ func addInputNamespacedScopes(allNamespaces map[string]schema.Scope, stage step.
 }
 
 // Adds the scope to the namespace map, as well as all resolved references from external namespaces.
-func addScopesWithReferences(allNamespaces map[string]schema.Scope, scope schema.Scope, prefix string) {
-	// First, just adds the scope
-	allNamespaces[prefix] = scope
+func addScopesWithReferences(allNamespaces map[string]map[string]*schema.ObjectSchema, scope schema.Scope, prefix string) {
+	// First, just adds the scope's objects.
+	allNamespaces[prefix] = scope.Objects()
 	// Next, checks all properties for resolved references that reference objects outside of this scope.
 	rootObject := scope.RootObject()
 	for propertyID, property := range rootObject.Properties() {
 		if property.Type().TypeID() == schema.TypeIDRef {
 			refProperty := property.Type().(schema.Ref)
-			if refProperty.Namespace() != schema.DEFAULT_NAMESPACE {
+			if refProperty.Namespace() != schema.SelfNamespace {
 				// Found a reference to an object that is not included in the scope. Add it to the map.
 				var referencedObject any = refProperty.GetObject()
 				refObjectSchema := referencedObject.(*schema.ObjectSchema)
-				allNamespaces[prefix+"."+propertyID] = schema.NewScopeSchema(refObjectSchema)
+				allNamespaces[prefix+"."+propertyID] = map[string]*schema.ObjectSchema{refObjectSchema.ID(): refObjectSchema}
 			}
 		}
 	}

--- a/workflow/executor_unit_test.go
+++ b/workflow/executor_unit_test.go
@@ -59,17 +59,17 @@ var testLifecycleStage = step.LifecycleStageWithSchema{
 }
 
 func TestAddOutputNamespacedScopes(t *testing.T) {
-	allNamespaces := make(map[string]schema.Scope)
+	allNamespaces := make(map[string]map[string]*schema.ObjectSchema)
 	namespacePrefix := "TEST_PREFIX_"
 	addOutputNamespacedScopes(allNamespaces, testLifecycleStage, namespacePrefix)
 	expectedNamespace := namespacePrefix + "outputC"
 	assert.Equals(t, len(allNamespaces), 1)
 	assert.MapContainsKey(t, expectedNamespace, allNamespaces)
-	assert.Equals(t, "testObjectC", allNamespaces[expectedNamespace].Root())
+	assert.MapContainsKey(t, "testObjectC", allNamespaces[expectedNamespace])
 }
 
 func TestAddInputNamespacedScopes(t *testing.T) {
-	allNamespaces := make(map[string]schema.Scope)
+	allNamespaces := make(map[string]map[string]*schema.ObjectSchema)
 	namespacePrefix := "TEST_PREFIX_"
 	expectedInputs := map[string]string{
 		"scopeA": "testObjectA",
@@ -81,12 +81,12 @@ func TestAddInputNamespacedScopes(t *testing.T) {
 	for expectedInput, expectedObject := range expectedInputs {
 		expectedNamespace := namespacePrefix + expectedInput
 		assert.MapContainsKey(t, expectedNamespace, allNamespaces)
-		assert.Equals(t, expectedObject, allNamespaces[expectedNamespace].Root())
+		assert.MapContainsKey(t, expectedObject, allNamespaces[expectedNamespace])
 	}
 }
 
 func TestAddScopesWithMissingCache(t *testing.T) {
-	allNamespaces := make(map[string]schema.Scope)
+	allNamespaces := make(map[string]map[string]*schema.ObjectSchema)
 	externalRef3 := schema.NewNamespacedRefSchema("scopeTestObjectC", "not-applied-namespace", nil)
 	notAppliedExternalRefProperty := schema.NewPropertySchema(
 		externalRef3,
@@ -117,7 +117,7 @@ func TestAddScopesWithMissingCache(t *testing.T) {
 
 func TestAddScopesWithReferences(t *testing.T) {
 	// Test that the scope itself and the resolved references are added.
-	allNamespaces := make(map[string]schema.Scope)
+	allNamespaces := make(map[string]map[string]*schema.ObjectSchema)
 	internalRef := schema.NewRefSchema("scopeTestObjectA", nil)
 	internalRefProperty := schema.NewPropertySchema(
 		internalRef,
@@ -178,9 +178,9 @@ func TestAddScopesWithReferences(t *testing.T) {
 			"scopeTestObjectB", map[string]*schema.PropertySchema{},
 		),
 	)
-	testScope.ApplyScope(scopeToApply, "test-namespace-1")
-	testScope.ApplyScope(scopeToApply, "$.test-namespace-2")
-	testScope.ApplyScope(testScope, schema.DEFAULT_NAMESPACE)
+	testScope.ApplyNamespace(scopeToApply.Objects(), "test-namespace-1")
+	testScope.ApplyNamespace(scopeToApply.Objects(), "$.test-namespace-2")
+	testScope.ApplySelf()
 	addScopesWithReferences(allNamespaces, testScope, "$")
 	assert.Equals(t, len(allNamespaces), 3)
 	assert.MapContainsKey(t, "$", allNamespaces)
@@ -229,17 +229,18 @@ func TestApplyAllNamespaces_Pass(t *testing.T) {
 	assert.Panics(t, func() {
 		ref2Schema.GetObject()
 	})
-	allNamespaces := make(map[string]schema.Scope)
-	allNamespaces["test-namespace-1"] = schema.NewScopeSchema(
-		schema.NewObjectSchema(
+	allNamespaces := make(map[string]map[string]*schema.ObjectSchema)
+	// Test one manual map creation, and one using scope's `Objects()` method.
+	allNamespaces["test-namespace-1"] = map[string]*schema.ObjectSchema{
+		"scopeTestObjectB": schema.NewObjectSchema(
 			"scopeTestObjectB", map[string]*schema.PropertySchema{},
 		),
-	)
+	}
 	allNamespaces["test-namespace-2"] = schema.NewScopeSchema(
 		schema.NewObjectSchema(
 			"scopeTestObjectC", map[string]*schema.PropertySchema{},
 		),
-	)
+	).Objects()
 	// Call the function under test and validate that all caches are set.
 	assert.NoError(t, applyAllNamespaces(allNamespaces, testScope))
 	assert.NoError(t, testScope.ValidateReferences())
@@ -289,12 +290,12 @@ func TestApplyAllNamespaces_MissingNamespace(t *testing.T) {
 		ref2Schema.GetObject()
 	})
 	// Only add test-namespace-1
-	allNamespaces := make(map[string]schema.Scope)
+	allNamespaces := make(map[string]map[string]*schema.ObjectSchema)
 	allNamespaces["test-namespace-1"] = schema.NewScopeSchema(
 		schema.NewObjectSchema(
 			"scopeTestObjectB", map[string]*schema.PropertySchema{},
 		),
-	)
+	).Objects()
 	err := applyAllNamespaces(allNamespaces, testScope)
 	assert.Error(t, err)
 	assert.Contains(t, err.Error(), "error validating references")
@@ -324,12 +325,12 @@ func TestApplyAllNamespaces_InvalidObject(t *testing.T) {
 			},
 		),
 	)
-	allNamespaces := make(map[string]schema.Scope)
+	allNamespaces := make(map[string]map[string]*schema.ObjectSchema)
 	allNamespaces["test-namespace-1"] = schema.NewScopeSchema(
 		schema.NewObjectSchema(
 			"scopeTestObjectB", map[string]*schema.PropertySchema{},
 		),
-	)
+	).Objects()
 	assert.PanicsContains(
 		t,
 		func() {

